apiserver/facades/agent/caasagent/caasagent.go

```diff
@@ -4,14 +4,173 @@
 package caasagent
 
 import (
+	"context"
+
+	"github.com/juju/names/v6"
+
 	"github.com/juju/juju/apiserver/common"
-	"github.com/juju/juju/apiserver/common/cloudspec"
-	"github.com/juju/juju/apiserver/common/model"
+	commonmodel "github.com/juju/juju/apiserver/common/model"
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/apiserver/internal"
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/core/crossmodel"
+	coremodel "github.com/juju/juju/core/model"
+	"github.com/juju/juju/core/network"
+	corewatcher "github.com/juju/juju/core/watcher"
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
 )
 
-// FacadeV2 is the V2 facade of the caas agent
+// StubService will be replaced once the implementation is finished.
+type StubService interface {
+	// CloudSpec returns the cloud spec for the model.
+	CloudSpec(ctx context.Context) (environscloudspec.CloudSpec, error)
+}
+
+// ModelService represents the credential service provided by the
+// provider.
+type ModelService interface {
+	WatchModelCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (corewatcher.NotifyWatcher, error)
+}
+
+// ControllerConfigService is an interface that provides the controller
+// configuration for the model.
+type ControllerConfigService interface {
+	ControllerConfig(context.Context) (controller.Config, error)
+}
+
+// ExternalControllerService defines the methods that the controller
+// facade needs from the controller state.
+type ExternalControllerService interface {
+	// ControllerForModel returns the controller record that's associated
+	// with the modelUUID.
+	ControllerForModel(ctx context.Context, modelUUID string) (*crossmodel.ControllerInfo, error)
+
+	// UpdateExternalController persists the input controller
+	// record.
+	UpdateExternalController(ctx context.Context, ec crossmodel.ControllerInfo) error
+}
+
+// ModelConfigService is an interface that provides access to the
+// model configuration.
+type ModelConfigService interface {
+	// ModelConfig returns the current config for the model.
+	ModelConfig(ctx context.Context) (*config.Config, error)
+	// Watch returns a watcher that returns keys for any changes to model
+	// config.
+	Watch() (corewatcher.StringsWatcher, error)
+}
+
+// ControllerConfigState defines the methods needed by
+// ControllerConfigAPI
+type ControllerConfigState interface {
```

Comment:
Would like to see comments on these functions and ideally the removal of `ModelExists` as this seems like a Mongo bleed that we don't need anymore?

Reply:
ModelExists is out side of this scope I can see that now. Please ignore.